File: nim/allergies/allergies.nim
import algorithm

type
  Allergies* = object of RootObj
    score*: int

const things = [
  "eggs",
  "peanuts",
  "shellfish",
  "strawberries",
  "tomatoes",
  "chocolate",
  "pollen",
  "cats",
]

proc lst*(allergies: Allergies): seq[string] =
  result = @[]
  var remainder = allergies.score
  for i, _ in things:
    let score = 1 shl i
    let diff = remainder - score
    if diff shr (i + 1) == remainder shr (i + 1):
      result.add(things[i])
      remainder = diff 

proc isAllergicTo*(allergies: Allergies, thing: string): bool =
  lst(allergies).contains(thing)

File: ocaml/allergies/allergies.ml
type allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats

let allergen_bit allergen =
    match allergen with
    | Eggs -> 1
    | Peanuts -> 2
    | Shellfish -> 3
    | Strawberries -> 4
    | Tomatoes -> 5
    | Chocolate -> 6
    | Pollen -> 7
    | Cats -> 8

let allergens = [
    Eggs;
    Peanuts;
    Shellfish;
    Strawberries;
    Tomatoes;
    Chocolate;
    Pollen;
    Cats;
]

let allergic_to score allergen =
    let bit = allergen_bit allergen in
    (score land (1 lsl (bit - 1))) > 0

let allergies score =
    allergens
    |> List.filter_map (fun a ->
        if allergic_to score a then
            Some a
        else
            None
    )
